File: node/src/connection/key.test.ts
import { describe, expect, test } from "bun:test";
import { BootstrapKeyPayload } from "../pb/servicebridge/v1/bootstrap";
import { parseBootstrapKey } from "./key";

function buildKey(
	keyID: Uint8Array,
	secret: Uint8Array,
	caDer: Uint8Array,
): string {
	const bytes = BootstrapKeyPayload.encode({
		keyId: Buffer.from(keyID),
		secret: Buffer.from(secret),
		caCertDer: Buffer.from(caDer),
	}).finish();
	return `sb.${Buffer.from(bytes).toString("base64url")}`;
}

const validKeyID = new Uint8Array(8).fill(1);
const validSecret = new Uint8Array(32).fill(2);
const validCA = new Uint8Array(200).fill(3);

describe("parseBootstrapKey", () => {
	test("parses valid key with embedded CA", () => {
		const key = buildKey(validKeyID, validSecret, validCA);
		const parsed = parseBootstrapKey(key);
		expect(parsed.keyID).toEqual(Buffer.from(validKeyID));
		expect(parsed.secret).toEqual(Buffer.from(validSecret));
		expect(parsed.caCertDer).toEqual(Buffer.from(validCA));
	});

	test("rejects missing sb. prefix", () => {
		const bytes = BootstrapKeyPayload.encode({
			keyId: Buffer.from(validKeyID),
			secret: Buffer.from(validSecret),
			caCertDer: Buffer.from(validCA),
		}).finish();
		const raw = Buffer.from(bytes).toString("base64url");
		expect(() => parseBootstrapKey(raw)).toThrow(/prefix/i);
	});

	test("rejects empty key_id field", () => {
		const bytes = BootstrapKeyPayload.encode({
			keyId: Buffer.alloc(0),
			secret: Buffer.from(validSecret),
			caCertDer: Buffer.from(validCA),
		}).finish();
		const raw = `sb.${Buffer.from(bytes).toString("base64url")}`;
		expect(() => parseBootstrapKey(raw)).toThrow(/key_id/i);
	});

	test("rejects empty secret field", () => {
		const bytes = BootstrapKeyPayload.encode({
			keyId: Buffer.from(validKeyID),
			secret: Buffer.alloc(0),
			caCertDer: Buffer.from(validCA),
		}).finish();
		const raw = `sb.${Buffer.from(bytes).toString("base64url")}`;
		expect(() => parseBootstrapKey(raw)).toThrow(/secret/i);
	});

	test("rejects empty ca_cert_der field", () => {
		const bytes = BootstrapKeyPayload.encode({
			keyId: Buffer.from(validKeyID),
			secret: Buffer.from(validSecret),
			caCertDer: Buffer.alloc(0),
		}).finish();
		const raw = `sb.${Buffer.from(bytes).toString("base64url")}`;
		expect(() => parseBootstrapKey(raw)).toThrow(/ca_cert_der/i);
	});

	test("rejects invalid base64url", () => {
		expect(() => parseBootstrapKey("sb.!!!invalid!!!")).toThrow();
	});
});

File: go/internal/connection/key_test.go
package connection_test

import (
	"bytes"
	"crypto/rand"
	"encoding/base64"
	"errors"
	"math/big"
	"testing"
	"testing/quick"

	"github.com/service-bridge/sdk/go/internal/connection"
	pb "github.com/service-bridge/sdk/go/internal/pb/servicebridge/v1"
	"google.golang.org/protobuf/proto"
)

// encodeKey mirrors connection.EncodeBootstrapKey in the runtime — the SDK only
// ever reads keys, so the encoder lives in the test.
func encodeKey(t *testing.T, keyID, secret, caDER []byte) string {
	t.Helper()
	raw, err := proto.Marshal(&pb.BootstrapKeyPayload{KeyId: keyID, Secret: secret, CaCertDer: caDER})
	if err != nil {
		t.Fatalf("marshal payload: %v", err)
	}
	return "sb." + base64.RawURLEncoding.EncodeToString(raw)
}

func randomBytes(t *testing.T, n int) []byte {
	t.Helper()
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		t.Fatalf("rand: %v", err)
	}
	return b
}

func TestParseBootstrapKeyRoundTrip(t *testing.T) {
	t.Parallel()
	ca := newTestCA(t)
	keyID := randomBytes(t, 8)
	secret := randomBytes(t, 32)

	key, err := connection.ParseBootstrapKey(encodeKey(t, keyID, secret, ca.der))
	if err != nil {
		t.Fatalf("ParseBootstrapKey: %v", err)
	}
	if !bytes.Equal(key.KeyID, keyID) {
		t.Errorf("KeyID: got %x want %x", key.KeyID, keyID)
	}
	if !bytes.Equal(key.Secret, secret) {
		t.Errorf("Secret: got %x want %x", key.Secret, secret)
	}
	if !bytes.Equal(key.CACertDER, ca.der) {
		t.Error("CACertDER mismatch")
	}
	if key.CACert == nil || !key.CACert.Equal(ca.cert) {
		t.Error("CACert is not the embedded CA")
	}
}

func TestParseBootstrapKeyPropertyRoundTrip(t *testing.T) {
	t.Parallel()
	ca := newTestCA(t)

	roundTrip := func(keyID [8]byte, secret [32]byte) bool {
		key, err := connection.ParseBootstrapKey(encodeKey(t, keyID[:], secret[:], ca.der))
		if err != nil {
			t.Logf("ParseBootstrapKey: %v", err)
			return false
		}
		return bytes.Equal(key.KeyID, keyID[:]) &&
			bytes.Equal(key.Secret, secret[:]) &&
			bytes.Equal(key.CACertDER, ca.der)
	}
	if err := quick.Check(roundTrip, &quick.Config{MaxCount: 200}); err != nil {
		t.Fatalf("round-trip property failed: %v", err)
	}
}

func TestParseBootstrapKeyRejects(t *testing.T) {
	t.Parallel()
	ca := newTestCA(t)
	valid := encodeKey(t, randomBytes(t, 8), randomBytes(t, 32), ca.der)

	tests := []struct {
		name string
		raw  string
	}{
		{"empty", ""},
		{"missing prefix", valid[len("sb."):]},
		{"wrong prefix", "sbx" + valid[len("sb."):]},
		{"broken base64", "sb.!!!not-base64!!!"},
		{"padded base64", "sb." + base64.StdEncoding.EncodeToString([]byte{0xff, 0xfe})},
		{"not a proto message", "sb." + base64.RawURLEncoding.EncodeToString([]byte{0xff, 0xff, 0xff, 0xff})},
		{"empty payload", encodeKey(t, nil, nil, nil)},
		{"key_id too short", encodeKey(t, randomBytes(t, 7), randomBytes(t, 32), ca.der)},
		{"key_id too long", encodeKey(t, randomBytes(t, 9), randomBytes(t, 32), ca.der)},
		{"secret too short", encodeKey(t, randomBytes(t, 8), randomBytes(t, 31), ca.der)},
		{"secret too long", encodeKey(t, randomBytes(t, 8), randomBytes(t, 33), ca.der)},
		{"empty ca", encodeKey(t, randomBytes(t, 8), randomBytes(t, 32), nil)},
		{"broken ca der", encodeKey(t, randomBytes(t, 8), randomBytes(t, 32), []byte{0x30, 0x03, 0x02, 0x01, 0x01})},
		{"ca der truncated", encodeKey(t, randomBytes(t, 8), randomBytes(t, 32), ca.der[:len(ca.der)/2])},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			_, err := connection.ParseBootstrapKey(tc.raw)
			if err == nil {
				t.Fatal("expected an error")
			}
			if !errors.Is(err, connection.ErrKey) {
				t.Errorf("kind: got %v want ErrKey", err)
			}
		})
	}
}

func TestParseBootstrapKeyRejectsRandomJunk(t *testing.T) {
	t.Parallel()
	// A random blob must never decode into a usable key. proto.Unmarshal accepts
	// some junk as an empty message, so the length checks carry the weight here.
	for range 200 {
		n, err := rand.Int(rand.Reader, big.NewInt(64))
		if err != nil {
			t.Fatalf("rand: %v", err)
		}
		blob := randomBytes(t, int(n.Int64()))
		raw := "sb." + base64.RawURLEncoding.EncodeToString(blob)
		if _, err := connection.ParseBootstrapKey(raw); err == nil {
			t.Fatalf("random blob parsed as a valid key: %x", blob)
		}
	}
}
